File: packages/ai-service/src/workflow/runner.ts
/**
 * runner — the single entry point for executing a workflow.
 * ==================================================================
 * Everything that can start a workflow — the Run button, a webhook, the
 * Automation Engine, a mission task through `workflow.run` — comes through
 * here. That is the point: there is exactly one place that decides what a
 * run IS, so there is exactly one place that can get versioning, scoping,
 * checkpointing or governance wrong.
 *
 * One run, in order:
 *
 *   1. resolve the workflow and the project
 *   2. freeze the graph as an immutable VERSION (or reuse the identical one)
 *   3. compute the version's authority ENVELOPE
 *   4. register the run's least-privilege SCOPE from that envelope
 *   5. create the durable RUN record
 *   6. build the GOVERNOR — the one door to the Capability Fabric
 *   7. execute, checkpointing after every node
 *   8. release the scope, always
 *
 * Step 4 is the one that is easy to skip and must not be: without it a
 * workflow runs with the machine's full authority regardless of what its
 * graph actually contains, and the envelope becomes decoration.
 */

import type { CapabilityFabric } from '@aura/capability-fabric';
import type { PipelineManager } from '../pipeline';
import type { RunScopeRegistry } from '../fabric/scopes';
import { secrets as secretStore } from '../secrets';
import { computeEnvelope, type AuthorityEnvelope } from './envelope';
import { runWorkflow as executeGraph, type RunOptions, type RunResult } from './engine';
import { createGovernor } from './governor';
import { createAgentRunner } from './agent/runner';
import type { AgentTrace } from './agent/types';
import { WorkflowRunStore } from './run/store';
import type { RunTrigger, WorkflowRun } from './run/types';
import { WorkflowVersionStore, type WorkflowVersion } from './versions';
import type { NodeIO, RunEvent, Workflow } from './types';

export interface RunnerDeps {
  pipeline: PipelineManager;
  fabric: CapabilityFabric | null;
  runScopes: RunScopeRegistry;
  versions: WorkflowVersionStore;
  runs: WorkflowRunStore;
}

export interface StartRunInput {
  workflow: Workflow;
  projectId: string;
  projectPath: string;
  projectName: string;
  trigger: RunTrigger;
  inputs?: Record<string, string>;
  /** Capabilities the human authorized for this run, if any. */
  approvedCapabilities?: string[];
  actor?: { kind: 'human' | 'agent' | 'system'; id: string };
  signal?: AbortSignal;
  timeoutMs?: number;
  /**
   * Called with the durable record the moment it exists, before the first
   * node runs. The caller needs the run id to register a cancel handle,
   * and it cannot wait for `start()` to resolve to get it — by then the
   * run is over and cancelling it is meaningless.
   */
  onRunCreated?: (run: WorkflowRun) => void;
}

export interface StartedRun {
  run: WorkflowRun;
  version: WorkflowVersion;
  envelope: AuthorityEnvelope;
  result: RunResult;
}

const DEFAULT_ACTOR = { kind: 'human' as const, id: 'user' };

export class WorkflowRunner {
  constructor(private readonly deps: RunnerDeps) {}

  /** What a workflow would be allowed to do, without running it. */
  envelopeOf(wf: Workflow): AuthorityEnvelope {
    return computeEnvelope(wf.nodes);
  }

  /**
   * Start a run.
   *
   * Returns once the run reaches a terminal state OR parks on an approval.
   * A parked run is not an error — `result.runState === 'awaiting-approval'`
   * and the record is durable and resumable.
   */
  async start(input: StartRunInput, emit: (e: RunEvent) => void): Promise<StartedRun> {
    const { workflow, projectId, projectPath } = input;
    const actor = input.actor ?? DEFAULT_ACTOR;

    const version = this.deps.versions.ensureVersionForRun(workflow, `run:${actor.id}`);
    const envelope = computeEnvelope(version.nodes);

    const run = this.deps.runs.create({
      workflowId: workflow.id,
      versionId: version.id,
      workflowName: version.name,
      projectId,
      projectPath,
      trigger: input.trigger,
      inputs: input.inputs,
    });
    input.onRunCreated?.(run);

    const result = await this.execute({ ...input, version, envelope, run, actor }, emit);
    return { run: this.deps.runs.get(workflow.id, run.id) ?? run, version, envelope, result };
  }

  /**
   * Pick up a parked or interrupted run where it stopped.
   *
   * Nodes that already succeeded are REPLAYED from their checkpointed
   * output rather than re-executed. That is the whole safety argument for
   * resume: a workflow that committed and pushed before parking must not
   * commit and push again when the human says yes.
   *
   * Refuses when the version is gone (nothing to execute) or when nothing
   * was checkpointed (nothing to resume from) — a resume that silently
   * restarted from the beginning would be the worst possible behaviour.
   */
  async resume(
    workflow: Workflow,
    runId: string,
    emit: (e: RunEvent) => void,
    opts: {
      approvedCapabilities?: string[];
      signal?: AbortSignal;
      actor?: StartRunInput['actor'];
      timeoutMs?: number;
      onRunCreated?: (run: WorkflowRun) => void;
    } = {},
  ): Promise<StartedRun | { error: string }> {
    const previous = this.deps.runs.get(workflow.id, runId);
    if (!previous) return { error: 'no such run' };
    if (previous.state === 'succeeded') return { error: 'that run already succeeded — there is nothing to resume' };
    if (!previous.resumable) return { error: previous.notResumableReason ?? 'that run cannot be resumed' };

    const version = this.deps.versions.get(workflow.id, previous.versionId);
    if (!version) {
      return { error: `the version this run executed (${previous.versionId}) no longer exists, so it cannot be resumed` };
    }

    /* Replay set: every node that banked a result. A node parked on an
       approval is deliberately NOT in it — that node is the one the human
       just answered, and it is what the resumed run exists to perform. */
    const replay: Record<string, NodeIO & { port?: string }> = {};
    const agentResume: Record<string, AgentTrace> = {};
    for (const node of Object.values(previous.nodes)) {
      if (node.state === 'succeeded' && node.output) {
        replay[node.nodeId] = { text: node.output.text, data: node.output.data, files: node.output.files, port: node.output.port };
        continue;
      }
      /* A PARKED agent is not replayed — it is continued. Its ledger
         carries the transcript, the iterations already spent and the wall
         clock already used, so the resumed agent keeps its reasoning and
         keeps its bounds. Replaying it would restart the budget; skipping
         it would lose the reasoning. */
      if (node.state === 'awaiting-approval' && node.agentTrace) {
        agentResume[node.nodeId] = node.agentTrace as AgentTrace;
      }
    }
    if (!Object.keys(replay).length && !Object.keys(agentResume).length && previous.state !== 'awaiting-approval') {
      return { error: 'no node completed in that run, so there is no checkpoint to resume from' };
    }

    const envelope = computeEnvelope(version.nodes);
    const run = this.deps.runs.create({
      workflowId: workflow.id,
      versionId: version.id,
      workflowName: version.name,
      projectId: previous.projectId,
      projectPath: previous.projectPath,
      trigger: { kind: 'resume', of: previous.id },
      inputs: previous.inputs,
    });
    run.vars = { ...previous.vars };
    opts.onRunCreated?.(run);

    const result = await this.execute({
      workflow: { ...workflow, nodes: version.nodes, edges: version.edges },
      projectId: previous.projectId,
      projectPath: previous.projectPath,
      projectName: workflow.name,
      trigger: run.trigger,
      inputs: previous.inputs,
      approvedCapabilities: opts.approvedCapabilities,
      actor: opts.actor ?? DEFAULT_ACTOR,
      signal: opts.signal,
      timeoutMs: opts.timeoutMs,
      version,
      envelope,
      run,
      replay,
      agentResume,
    }, emit);

    return { run: this.deps.runs.get(workflow.id, run.id) ?? run, version, envelope, result };
  }

  /** The shared execution body. Registers the scope; always releases it. */
  private async execute(
    input: StartRunInput & {
      version: WorkflowVersion;
      envelope: AuthorityEnvelope;
      run: WorkflowRun;
      actor: NonNullable<StartRunInput['actor']>;
      replay?: Record<string, NodeIO & { port?: string }>;
      agentResume?: Record<string, AgentTrace>;
    },
    emit: (e: RunEvent) => void,
  ): Promise<RunResult> {
    const { run, version, envelope, actor } = input;

    // Least privilege, derived from the graph that is actually executing.
    const scopes = envelope.scopes.map((s) => s.scope);
    this.deps.runScopes.register(run.id, scopes);

    const governor = this.deps.fabric
      ? createGovernor({
          fabric: this.deps.fabric,
          secrets: secretStore,
          workflowId: run.workflowId,
          runId: run.id,
          projectId: run.projectId,
          projectPath: run.projectPath,
          actor,
          approvedCapabilities: input.approvedCapabilities,
        })
      : undefined;

    /* The agent runtime, built from the SAME envelope the governor uses.
       That shared envelope is what makes "the agent's tools ≤ the
       workflow's permissions" structural rather than advisory: both doors
       into the Fabric are bounded by one authority, computed once from the
       version that is actually executing. */
    const agents = this.deps.fabric
      ? createAgentRunner({
          fabric: this.deps.fabric,
          pipeline: this.deps.pipeline,
          envelope,
          workflowId: run.workflowId,
          runId: run.id,
          projectId: run.projectId,
          projectPath: run.projectPath,
          redact: governor?.redact ?? ((t: string) => t),
          signal: input.signal,
        })
      : undefined;

    const options: RunOptions = {
      projectId: run.projectId,
      projectPath: run.projectPath,
      projectName: input.projectName,
      pipeline: this.deps.pipeline,
      inputs: run.inputs,
      signal: input.signal,
      run,
      runs: this.deps.runs,
      governor,
      agents,
      timeoutMs: input.timeoutMs,
      replay: input.replay,
      agentResume: input.agentResume,
    };

    try {
      return await executeGraph({ ...input.workflow, nodes: version.nodes, edges: version.edges }, options, emit);
    } finally {
      // A run's authority never outlives the run, on any exit path.
      this.deps.runScopes.release(run.id);
    }
  }
}

File: backend/aura/workflow/runner.py
"""WorkflowRunner — port of runner.ts. THE single canonical entry point.

Manual/API, scheduler, automation and (later) Central Agent ALL call
`WorkflowRunner.start` / `.resume`. There is no other execution path.
"""
from __future__ import annotations

import asyncio
from typing import Any

from ..persistence.runs import WorkflowRunStore
from ..persistence.versions import WorkflowVersionStore
from .engine import run_workflow
from .governor import create_governor


def compute_envelope(nodes: list[dict]) -> dict:
    """Minimal envelope: scopes actually required by the executing graph.

    Full TS envelope fields (hosts/notRequested/cannot) are presentation
    surface frozen for Phase 10 wire compat; governance depends on the
    scope list, which must be exact.
    """
    from .nodes_core import GOVERNED_TYPES

    scopes: dict[str, dict] = {}
    for n in nodes:
        if n["type"] not in GOVERNED_TYPES:
            continue
        binding_cap = {
            "shell-command": "terminal.execute", "export-file": "filesystem.write",
            "git-status": "git.status", "changed-files": "git.status",
            "git-diff": "git.diff", "git-commit": "git.commit",
            "git-branch": "git.branch",
            "http-request": "http.request",
        }.get(n["type"])
        cap = next((c for c in _manifest_caps() if c["id"] == binding_cap), None)
        if not cap:
            continue
        entry = scopes.setdefault(cap["id"], {"scope-set": set(), "capabilityIds": [], "risk": cap["risk"]})
        if n["id"] not in [x for x in entry["capabilityIds"]]:
            entry["capabilityIds"].append(n["id"])
        for p in cap.get("permissions") or []:
            entry["scope-set"].add(p)
    return {"capabilities": sorted(
        [{"capabilityId": cid, "scopes": sorted(v["scope-set"]),
          "risk": v["risk"], "nodeIds": v["capabilityIds"]}
         for cid, v in scopes.items()], key=lambda x: x["capabilityId"])}


_MANIFEST_CACHE = None


def _manifest_caps():
    global _MANIFEST_CACHE
    if _MANIFEST_CACHE is None:
        import json
        from pathlib import Path

        f = Path(__file__).parent.parent / "fabric" / "manifest.json"
        _MANIFEST_CACHE = json.loads(f.read_text())["capabilities"]
    return _MANIFEST_CACHE


class WorkflowRunner:
    """THE canonical execution path."""

    def __init__(self, *, fabric, run_scopes, versions: WorkflowVersionStore,
                 runs: WorkflowRunStore, secrets=None, pipeline=None,
                 projects=None) -> None:
        self.fabric = fabric
        self.run_scopes = run_scopes
        self.versions = versions
        self.runs = runs
        self.secrets = secrets
        self.pipeline = pipeline
        self.projects = projects
        self.live_runs: dict[str, asyncio.Event] = {}

    # ── convergence seam ────────────────────────────────────────────────
    # manual/API, scheduler, automation and Central Agent all funnel here.
    async def start_workflow_run(self, workflow: dict, *, project_id: str,
                                 project_path: str, project_name: str = "",
                                 trigger: dict, inputs: dict | None = None,
                                 approved_capabilities: list[str] | None = None,
                                 actor: dict | None = None,
                                 emit: Any = lambda e: None,
                                 on_run_created=None,
                                 max_node_executions: int | None = None) -> dict:
        return await self.start({"workflow": workflow, "projectId": project_id,
                                 "projectPath": project_path,
                                 "projectName": project_name or workflow.get("name", ""),
                                 "trigger": trigger, "inputs": inputs,
                                 "approvedCapabilities": approved_capabilities,
                                 "actor": actor,
                                 "maxNodeExecutions": max_node_executions}, emit)

    workflows = None  # optional store seam; automation resolves ids via it

    def _agent_runner(self, *, run: dict, envelope: dict, redact, signal):
        if self.fabric is None:
            return None
        from .agent.runner import AgentRunner

        return AgentRunner(
            fabric=self.fabric, envelope=envelope, redact=redact,
            workflow_id=run["workflowId"], run_id=run["id"],
            project_id=run["projectId"], project_path=run["projectPath"],
            model=getattr(self, "model", None),
            actor={"kind": "agent", "id": "agent:workflow"})

    def cancel_workflow_run(self, run_id: str) -> bool:
        ev = self.live_runs.get(run_id)
        if not ev:
            return False
        ev.set()          # AbortSignal parity
        return True

    async def start(self, input: dict, emit) -> dict:
        workflow = input["workflow"]
        actor = input.get("actor") or {"kind": "human", "id": "user"}
        version = self.versions.ensure_version_for_run(workflow, f"run:{actor['id']}")
        envelope = compute_envelope(version["nodes"])

        run = self.runs.create({
            "workflowId": workflow["id"], "versionId": version["id"],
            "workflowName": version["name"], "projectId": input["projectId"],
            "projectPath": input["projectPath"], "trigger": input["trigger"],
            "inputs": input.get("inputs"),
        })
        cancel_event = asyncio.Event()
        self.live_runs[run["id"]] = cancel_event
        if input.get("onRunCreated"):
            input["onRunCreated"](run)

        result = await self._execute(input, version, envelope, run, actor, cancel_event, emit)
        self.live_runs.pop(run["id"], None)
        fresh = self.runs.get(workflow["id"], run["id"])
        return {"run": fresh or run, "version": version, "envelope": envelope,
                "result": result}

    async def resume(self, workflow: dict, run_id: str, emit, opts: dict | None = None):
        opts = opts or {}
        previous = self.runs.get(workflow["id"], run_id)
        if not previous:
            return {"error": "no such run"}
        if previous["state"] == "succeeded":
            return {"error": "that run already succeeded — there is nothing to resume"}
        if previous.get("supersededBy"):
            return {"error": f"that run was already continued as run {previous['supersededBy']}"}
        if not previous.get("resumable"):
            return {"error": previous.get("notResumableReason") or "that run cannot be resumed"}

        version = self.versions.get(workflow["id"], previous["versionId"])
        if not version:
            return {"error": f"the version this run executed ({previous['versionId']}) no longer exists, so it cannot be resumed"}

        replay: dict[str, dict] = {}
        agent_resume: dict[str, dict] = {}
        for node in previous["nodes"].values():
            if node["state"] == "succeeded" and node.get("output"):
                replay[node["nodeId"]] = {"text": node["output"].get("text"),
                                          "data": node["output"].get("data"),
                                          "files": node["output"].get("files"),
                                          "port": node["output"].get("port")}
            elif node["state"] == "awaiting-approval" and node.get("agentTrace"):
                agent_resume[node["nodeId"]] = node["agentTrace"]
        if not replay and not agent_resume and previous["state"] != "awaiting-approval":
            return {"error": "no node completed in that run, so there is no checkpoint to resume from"}

        envelope = compute_envelope(version["nodes"])
        run = self.runs.create({
            "workflowId": workflow["id"], "versionId": version["id"],
            "workflowName": version["name"], "projectId": previous["projectId"],
            "projectPath": previous["projectPath"],
            "trigger": {"kind": "resume", "of": previous["id"]},
            "inputs": previous.get("inputs"),
        })
        run["vars"] = dict(previous.get("vars") or {})
        cancel_event = asyncio.Event()
        self.live_runs[run["id"]] = cancel_event
        self.runs.mark_superseded(workflow["id"], previous["id"], run["id"])
        if opts.get("onRunCreated"):
            opts["onRunCreated"](run)

        result = await self._execute(
            {"workflow": {**workflow, "nodes": version["nodes"], "edges": version["edges"]},
             "projectId": previous["projectId"], "projectPath": previous["projectPath"],
             "projectName": workflow.get("name", ""), "trigger": run["trigger"],
             "inputs": previous.get("inputs"),
             "approvedCapabilities": opts.get("approvedCapabilities"),
             "actor": opts.get("actor") or {"kind": "human", "id": "user"},
             "replay": replay, "agentResume": agent_resume},
            version, envelope, run, opts.get("actor") or {"kind": "human", "id": "user"},
            cancel_event, emit)
        self.live_runs.pop(run["id"], None)
        fresh = self.runs.get(workflow["id"], run["id"])
        return {"run": fresh or run, "version": version, "envelope": envelope,
                "result": result}

    async def _execute(self, input: dict, version: dict, envelope: dict,
                       run: dict, actor: dict, cancel_event: asyncio.Event,
                       emit) -> dict:
        scopes: list[str] = []
        for c in envelope.get("capabilities") or []:
            scopes.extend(c.get("scopes") or [])
        self.run_scopes.register(run["id"], set(scopes))
        governor = None
        if self.fabric is not None:
            governor = create_governor(
                fabric=self.fabric,
                secrets=self.secrets or type("S", (), {"known_values": staticmethod(list)})(),
                workflow_id=run["workflowId"], run_id=run["id"],
                project_id=run["projectId"], project_path=run["projectPath"],
                actor=actor,
                approved_capabilities=input.get("approvedCapabilities"),
            )
        signal = {"aborted": False, "event": cancel_event}
        watcher = asyncio.ensure_future(_watch(cancel_event, signal))

        wf_exec = {"id": input["workflow"]["id"], "name": input["workflow"].get("name", ""),
                   "nodes": version["nodes"], "edges": version["edges"]}
        try:
            return await run_workflow(wf_exec, {
                "projectId": run["projectId"], "projectPath": run["projectPath"],
                "projectName": input.get("projectName"), "inputs": run.get("inputs"),
                "signal": signal, "run": run, "runs": self.runs,
                "governor": governor,
                "model": getattr(self, "model", None),
                "agents": self._agent_runner(run=run, envelope=envelope,
                                             redact=governor.redact if governor else (lambda t: t),
                                             signal=signal),
                "timeoutMs": input.get("timeoutMs"),
                "replay": input.get("replay"),
                "agentResume": input.get("agentResume"),
                "maxNodeExecutions": input.get("maxNodeExecutions"),
            }, emit)
        finally:
            watcher.cancel()
            self.run_scopes.remove(run["id"])


async def _watch(cancel_event: asyncio.Event, signal: dict) -> None:
    await cancel_event.wait()
    signal["aborted"] = True
